### PR TITLE
Fix InMemoryStore update issue

### PR DESCRIPTION
**Pull request by varunpuranik**
GetIndex method in InMemoryStore creates a snapshot of the data. If an update operation takes place during the creation of the snapshot, an exception is thrown. 

To fix the issue, remove the snapshot from the GetIndex method. 
Add locking around all the read/update operations .

### CODE REVIEW

File: edgelet/edgelet-http/src/unix.rs [hunk 1]
```diff
@@ -0,0 +1,91 @@
+// Copyright (c) Microsoft. All rights reserved.
```
Comment (dsajanice) [inline edgelet/edgelet-http/src/unix.rs]: Why are these changes required?

Comment (varunpuranik) [reply edgelet/edgelet-http/src/unix.rs]: Oops.. need to rebase, I suppose.. certainly not part of my changes

File: edge-util/src/Microsoft.Azure.Devices.Edge.Storage/InMemoryDbStore.cs [hunk 2]
```diff
@@ -13,106 +15,192 @@ namespace Microsoft.Azure.Devices.Edge.Storage
     /// </summary>
     class InMemoryDbStore : IDbStore
     {
-        // Deliberately using list (instead of a synchronied list like Concurrent dictionary) to 
-        // better simulate acutal key-value stores like RocksDb.
-        readonly List<(byte[], byte[])> keyValues;
+        readonly ItemKeyedCollection keyValues;
```
Comment (dsajanice) [inline edge-util/src/Microsoft.Azure.Devices.Edge.Storage/InMemoryDbStore.cs]: Curious - What is the reason to change from List to ItemKeyedCollection?

Comment (varunpuranik) [reply edge-util/src/Microsoft.Azure.Devices.Edge.Storage/InMemoryDbStore.cs]: Performance - KeyedCollection provides near O(1) performance for sequential access and indexed access.

Comment (dsajanice) [reply edge-util/src/Microsoft.Azure.Devices.Edge.Storage/InMemoryDbStore.cs]: There was a comment in there saying we wanted a list because it more closely simulated rocks db. Is that not required anymore?

Comment (varunpuranik) [reply edge-util/src/Microsoft.Azure.Devices.Edge.Storage/InMemoryDbStore.cs]: I would like that, but there are problems with a list that RocksDb does not have. So at this point I wanted something as close to RocksDB without requiring it to be identical.
